### ISSUE TITLE
ch 6.4.2 p 164 ptb_iterator丢失数据？

### ISSUE THREAD

**Issue by arisliang** [tensorlayer/chinese-book#19]
书中给出类似 https://tensorlayer.readthedocs.io/en/latest/modules/iterate.html?highlight=ptb_iterator#tensorlayer.iterate.ptb_iterator 的例子。
当batch_size=2, num_steps=3时，结果是:
>  print(x, y)
... [[ 0  1  2] <---x                       1st subset/ iteration
...  [10 11 12]]
... [[ 1  2  3] <---y
...  [11 12 13]]
...
... [[ 3  4  5]  <--- 1st batch input       2nd subset/ iteration
...  [13 14 15]] <--- 2nd batch input
... [[ 4  5  6]  <--- 1st batch target
...  [14 15 16]] <--- 2nd batch target
...
... [[ 6  7  8]                             3rd subset/ iteration
...  [16 17 18]]
... [[ 7  8  9]
...  [17 18 19]]

为什么(x, y)的pair直接从1st subset ([0, 1, 2], [1, 2, 3])跳到 2nd subset ([3, 4, 5], [4, 5, 6])，中间的([1, 2, 3], [2, 3, 4])和([2, 3, 4], [3, 4, 5])呢？
我尝试使用不同的batch_size和num_steps，还是没有办法得到中间跳过的subset.
比如batch_size=1, num_steps=3, 结果为:
> x: [[0 1 2]]
y: [[1 2 3]]
x: [[3 4 5]]
y: [[4 5 6]]
x: [[6 7 8]]
y: [[7 8 9]]
x: [[ 9 10 11]]
y: [[10 11 12]]
x: [[12 13 14]]
y: [[13 14 15]]
x: [[15 16 17]]
y: [[16 17 18]]

书中说batch_size设为1，所有的样本就会被一个一个放入网络。但其实即使batch_size为1，中间还是有数据被跳过和丢失了（比如x:[1, 2, 3], y: [2, 3, 4]).
看起来问题在于num_steps既控制window_size, 又控制stride，而像cnn的kernel里面这是由两个参数一起控制的。

**Comment by zsdonghao** [comment]
多少个batch就是把数据分为多少分，然后num_steps就是，RNN输入sequence的长度。
所以第一个batch就是：
x    --> y
0, 1, 2 --> 1, 2, 3
10, 11, 12 --> 11, 12, 13

**Comment by arisliang** [comment]
那第二个batch为什么不是:
> 1, 2, 3 --> 2, 3, 4
11, 12, 13 --> 12, 13, 14

**Comment by zsdonghao** [comment]
0, 1, 2 之后是 3, 4, 5 。state传过去，你自己体会一下。

**Comment by arisliang** [comment]
什么的state？rnn state? 如果是数据的iterator，不应该有具体的模型有关吧？因为数据可能会被用在不同模型上。

不知道从哪里开始去理解, 为什么(0, 1, 2)之后不是(1, 2, 3)，而是(3, 4, 5)。既然是整数，为什么0后面不是1，而是3.

**Comment by zsdonghao** [comment]
不是的，这个迭代和mnist的迭代不一样。
理论上来说，我们用 0, 1, 2,.....N  --> 1, 2, ----> N-1 来训练是最好的，但因为内存不够，我们才把它切开。

0, 1, 2 - 3, 4, 5 - 6, 7, 8 - .....
1, 2, 3 - 4, 5, 6 - 7, 8, 9 - ....

`-` 代表一段sequence_length之后的rnn state传输

**Comment by arisliang** [comment]
内存够不够可能取决于数据集，也取决于机器。不同问题和不同用户可能结论不同。
如果数据集大的话，可能内存不够，这时可以设置一个切开的参数；如果是小的数据集，那可能不用切开也是够用的。这个感觉由用户来决定会比较好。

**Comment by zsdonghao** [comment]
ptb数据那么长，不用mini-batch根本没发训练的。
大数据时代，几乎所有用到深度学习的问题，都没法不用mini-batch来训练。。

**Comment by arisliang** [comment]
哦，我不是说不用mini-batch，当然是要切成小的batch。我的意思是切法的问题。以上面的batch_size=1, num_steps=3为例。
当前的结果是：
batch 1:
x --> y: [[**0** 1 2]] --> [[1 2 3]]
batch 2:
x --> y: [[**3** 4 5]] --> [[4 5 6]]
这中间跳过了两个可能的batch
x --?> y: [[**1** 2 3]] --> [[2 3 4]]
x --?> y: [[**2** 3 4]] --> [[3 4 5]]

**Comment by zsdonghao** [comment]
@arisliang 你还是没理解。。上一个iteration和下一个iteration是连起来的，0-1-2-3-4-5 ，每一步的rnn state都会传输过去，所以根本没用你说的那2个batch。

**Comment by arisliang** [comment]
可以这样连的吗？我理解是用x来预测y，这样没有[1 2 3]，怎么预测[2 3 4]?
第一个iteration，用[0 1 2]预测[1 2 3]
第二个iteration，用[3 4 5]预测[4 5 6]
那用[1 2 3]来预测[2 3 4]的步骤是在哪里？

**Comment by zsdonghao** [comment]
这样的，rnn每一个输出都是基于之前的state和输入决定的。当学习了0～6，自然就学会了输入123输出234了，这个和我们用什么sequence length无关（这个只是看我们希望每次更新基于多长的数据而已）。具一个极端例子，sequence length=1，他依然能学会输入123时输出234的。

**Comment by arisliang** [comment]
你的意思是学了012->123和345->456，就自然学会了123->234和234->345? 

但是123->234和234->345从来都没有在数据对中训练，也就不会在loss中去调整。

我尝试找keras的例子，https://github.com/keras-team/keras/blob/master/examples/lstm_text_generation.py line 36-42. maxlen (相当于这里的num_steps), step来控制前进的步数，比如x从[0 1 2]到[3 4 5]，step就是3，这个是用户可以控制的，比如希望x从[0 1 2]到[5 6 7]，那么step就是5. 或者想要x生成[1 2 3]，那么step就是1.

**Comment by zsdonghao** [comment]
the final state of 012 already contain the information of 012, if you feed 123 after it, It will become 012123. The step/sequence length is just for defining “how many steps do we use to compute the loss for every update”. That is why for inferencing, we can feed in 1 word everytime. If you still confuse, you can contact me via WeChat, it is easier for me.

**Comment by arisliang** [comment]
It's like when using time-series prediction, one use (t0, t1, t2) to predict t3; and the next time window is to use (t1, t2, t3) to predict t4, rather than jumping to (t3, t4, t5) to predict t6.

If combining x, it becomes 012123 (overlapped, then combined y would become 123234, overlapped as well). But it's fine, because the artificially overlapped pair (012123, 123234) is never merged as so, and will never be fed to the network as a training pair; and (012123) should not be fed to network for meaningful prediction (but if someone feed it for fun anyway, that's a different story:)).

For prediction, we may feed 012345 together, or 0, 1, 2, 3, 4, 5 individually as wanted, because of the variable-sequence, like you said.

Yeah, i'm still confused. Will contact you via wechat.

**Comment by arisliang** [comment]
I see what you mean now.
Think I was misled by the thing "batch". The batch in this context of RNN is completely different from non-RNN (like MNIST).

In non-RNN, every batch is independent training data regardless of where and when the training happens.

In the RNN context however, because one batch's final state transferred to next batch's initial state within one training epoch, all batch concatenated as "one" training data in that epoch (the parameter may get updated depending on num_step, but the information gets carried from first batch until last batch).

Would like to suggest to use a different name other than "batch" to avoid confusion for beginners, as "batch" has this very different meaning in non-RNN context. Segment is a good one (have seen it mentioned in the documentation), just to emphasize the difference.

Back to the ptb_iterator. While its current behavior (non-overlapping, non-skipping) fits the RNN training well, it won't work for non-RNN needs.

In this sense, I still think it's good to have a separate parameter to help control the overlapping behavior as a more general data prepro tool. To still support RNN first, just need to make the parameter's default behavior as it is. For overlapping behavior needs, user can tweak the parameter as needed.

**Comment by zsdonghao** [comment]
@arisliang  hi，这个问题你完全搞懂了吗？能不能把你之前画的那张图贴到这里？我觉得那图解析的非常好～ 之后书里还能加上去。。

**Comment by arisliang** [comment]
感觉好像大概是搞懂了.
这个是当时和你讨论画的理解图 (synced many to many).
![ca77cb83cc0d813e26e0b3aeb2c75d0f](https://user-images.githubusercontent.com/1115503/38462498-bba59d1c-3b1a-11e8-8766-5c167b6569d1.jpg)

又看了一下书, 发现书中有一句话挺关键, 开始时忽略了. p160 ch6.4.1,
> 输入 "I love developing deep learning" 以预测 "love developing deep learning applications"

这句话明显指向了synched many to many的类型, 如果能把这句话画成图, 就更清楚了.

**Comment by zsdonghao** [comment]
@arisliang 收到～
